packages/babel-parser/src/parser/statement.js

```diff
@@ -539,13 +539,30 @@ export default class StatementParser extends ExpressionParser {
       return this.parseFor(node, init);
     }
 
+    // Check for `for (async of`, which is forbidden if this turns out to be
+    // a for-of loop.
+    const startsWithAsyncOf =
+      // `for await (async of []);` is allowed.
+      !(awaitAt > -1) &&
+      !startsWithLet &&
+      this.isContextual("async") &&
+      // This lookahead only occurs after `for (async`, which should be rare.
+      this.isLookaheadContextual("of");
+
     const refExpressionErrors = new ExpressionErrors();
     const init = this.parseExpression(true, refExpressionErrors);
     const isForOf = this.isContextual("of");
-    if (isForOf || this.match(tt._in)) {
-      if (isForOf && startsWithLet) {
+    if (isForOf) {
+      // Check for leading tokens that are forbidden in for-of loops:
+      if (startsWithLet) {
         this.raise(init.start, Errors.ForOfLet);
+      } else if (startsWithAsyncOf && init.type !== "ArrowFunctionExpression") {
```

Comment:
I don't think we need lookahead before, instead we can check if `awaitAt === -1`, `init.type` is `Identifier`, `init.name` is `async` and `!containsEsc` here.

Reply:
This doesn't quite work, because it rejects `for ((async) of []);` which should be legal.

Reply:
Though if I also do an early check for `tt.name`, that should be good enough? I'll try it out.

Reply:
Make sense. Well in that case we can check `type = Identifier` here and avoid the `of` lookahead.